File: leetcode/solutions/0701.InsertIntoABinarySearchTree/insertIntoBST.ts
import { BinaryTreeNode } from "../../../data_structures/BinaryTreeNode.ts";

export default function insertIntoBST(
  root: BinaryTreeNode<number> | null,
  val: number
): BinaryTreeNode<number> | null {
  if (!root) {
    return null;
  }

  let node = root;

  while (true) {
    if (node.val > val) {
      if (node.left) {
        node = node.left;
        continue;
      } else {
        node.left = { val, left: null, right: null };
        break;
      }
    } else {
      if (node.right) {
        node = node.right;
        continue;
      } else {
        node.right = { val, left: null, right: null };
        break;
      }
    }
  }

  return root;
}

File: leetcode/solutions/0701.InsertIntoABinarySearchTree/insertIntoBST.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func insertIntoBST(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return &TreeNode{Val: val}
	}

	if root.Val > val {
		if root.Left == nil {
			root.Left = &TreeNode{Val: val}
		} else {
			insertIntoBST(root.Left, val)
		}
	} else {
		if root.Right == nil {
			root.Right = &TreeNode{Val: val}
		} else {
			insertIntoBST(root.Right, val)
		}
	}

	return root
}
